src/gz/write.rs

```diff
@@ -180,3 +182,327 @@ impl<W: Write> Drop for GzEncoder<W> {
+            return Err(corrupt());
+        }
+        Ok(())
+    }
+
+    fn write_buf(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let (n, status) = self.inner.write_with_status(buf)?;
+
+        if status == Status::StreamEnd {
+            if n < buf.len() && self.crc_bytes.len() < 8 {
+                let remaining = buf.len() - n;
+                let crc_bytes = cmp::min(remaining, CRC_BYTES_LEN - self.crc_bytes.len());
+                self.crc_bytes.extend(&buf[n..n + crc_bytes]);
+                return Ok(n + crc_bytes);
+            }
+        }
+        Ok(n)
+    }
+}
+
+struct Wrapper<'a> {
+    buf: &'a [u8],
+    header_buf: &'a mut Vec<u8>,
+    pos: usize,
+}
+
+impl<'a> Wrapper<'a> {
+    fn buf_pos(&self) -> usize {
+        self.pos - self.header_buf.len()
+    }
+}
+
+impl<'a> Read for Wrapper<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let mut buf_pos = 0;
+
+        // read from header buffer
+        if self.pos < self.header_buf.len() {
+            buf_pos = cmp::min(buf.len(), self.header_buf.len() - self.pos);
+            (&mut buf[..buf_pos]).copy_from_slice(&self.header_buf[self.pos..self.pos + buf_pos]);
+            self.pos += buf_pos;
+        }
+
+        // read from current buf
+        if buf_pos < buf.len() && self.pos >= self.header_buf.len() {
+            let diff = self.pos - self.header_buf.len();
+            let size = cmp::min(buf.len() - buf_pos, self.buf.len() - diff);
+            (&mut buf[buf_pos..buf_pos + size]).copy_from_slice(&self.buf[..size]);
+            self.pos += size;
+            buf_pos += size;
+        }
+
+        Ok(buf_pos)
+    }
+}
+
+impl<W: Write> Write for GzDecoder<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if self.header.is_none() {
+            // trying to avoid buffer usage
+            let (res, pos) = if self.header_buf.is_empty() {
+                let mut cur = io::Cursor::new(buf);
+                let res = read_gz_header(&mut cur);
+                (res, cur.position() as usize)
+            } else {
+                let mut wrp = Wrapper {
+                    buf: buf,
+                    header_buf: &mut self.header_buf,
+                    pos: 0,
+                };
+                let res = read_gz_header(&mut wrp);
+                (res, wrp.buf_pos())
+            };
+
+            match res {
+                Err(err) => {
+                    if err.kind() == io::ErrorKind::UnexpectedEof {
+                        // not enough data for header, save to the buffer
+                        self.header_buf.extend(buf);
+                        Ok(buf.len())
+                    } else {
+                        Err(err)
+                    }
+                }
+                Ok(header) => {
+                    self.header = Some(header);
+                    self.header_buf.truncate(0);
+                    let n = self.write_buf(&buf[pos..])?;
```

Comment:
Since bytes have been "committed" at this point in the sense that the header has been read, we shouldn't continue to call `write_buf` and instead report a short write was successful (that way it'll act atomically)

Reply:
fixed